packages/core/core-v2/src/Parcel.js

```diff
@@ -0,0 +1,198 @@
+// @flow
+'use strict';
+import { AbortController } from 'abortcontroller-polyfill/dist/cjs-ponyfill';
+import Watcher from '@parcel/watcher';
+import PQueue from 'p-queue';
+import AssetGraph from './AssetGraph';
+import { Node } from './Graph';
+import type { Dependency, Asset, File } from './types';
+import TransformerRunner from './TransformerRunner';
+import ResolverRunner from './ResolverRunner';
+import BundlerRunner from './BundlerRunner';
+import PackagerRunner from './PackagerRunner';
+
+// TODO: use custom config if present
+const defaultConfig = require('@parcel/config-default');
+
+const abortError = new Error('Build aborted');
+
+type CliOpts = {
+  watch?: boolean
+}
+
+type ParcelOpts = {
+  entries: Array<string>,
+  cwd?: string,
+  cliOpts: CliOpts,
+}
+
+type Signal = {
+  aborted: boolean,
+  addEventListener?: Function,
+}
+
+type BuildOpts = {
+  signal: Signal,
+  shallow?: boolean,
+}
+
+export default class Parcel {
+  entries: Array<string>;
+  rootDir: string;
+  graph: AssetGraph;
+  watcher: Watcher;
+  updateQueue: PQueue;
+  mainQueue: PQueue;
+  transformerRunner: TransformerRunner;
+  resolverRunner: ResolverRunner;
+  bundlerRunner: BundlerRunner;
+  packagerRunner: PackagerRunner;
+
+  constructor({
+    entries,
+    cliOpts = {},
+  }: ParcelOpts) {
+    this.rootDir = process.cwd();
+
+    this.graph = new AssetGraph({ entries, rootDir: this.rootDir });
+    this.watcher = cliOpts.watch ? new Watcher() : null;
+    this.updateQueue = new PQueue({ autoStart: false });
+    this.mainQueue = new PQueue({ autoStart: false });
+
+    this.transformerRunner = new TransformerRunner({ parcelConfig: defaultConfig, cliOpts });
+    this.resolverRunner = new ResolverRunner();
+    this.bundlerRunner = new BundlerRunner({ parcelConfig: defaultConfig, cliOpts });
+    this.packagerRunner = new PackagerRunner({ parcelConfig: defaultConfig, cliOpts });
+  }
+
+  async run() {
+    let controller = new AbortController();
+    let signal = controller.signal;
+
+    let buildPromise = this.build({ signal });
+
+    if (this.watcher) {
+      this.watcher.on('change', event => {
+        // The build step starts by updating the graph with all of the changes in the update queue
+        if (!this.updateQueue.isPaused) {
+          // If a change comes in while the graph is being updated, just add it to the update queue
+          this.handleChange(event);
+        } else {
+          // Otherwise cancel any work that may be in progress, add the change to the update queue,
+          // and start a new build.
+          controller.abort();
+          this.mainQueue.pause();
+          this.mainQueue.clear();
```

Comment:
What about aborting things in the update queue? If multiple subsequent change events came in for a file, the earlier ones wouldn't get canceled and replaced with the new ones, they'd pile up.

Better question: why are there two queues? Why not just have a single one?

Reply:
I'm seeing a lot of questions related to cancellation, so I am just going to explain my vision here and that should help clear up a lot of questions at once. (Suggestions for better names always accepted) My graph building implementation makes a distinction between two phases, updating the graph and completing the graph. During the `update` phase, the graph is pruned from the middle. If the graph is in an inconsistent state with the file system, anything that is no longer correct will be cleaned up. During the `complete` phase, the graph is filled out from the edges. Any files that have not been transformed will be transformed and any dependencies that have not been resolved will be resolved until there is nothing left to transform or resolve. Now if a filesystem change comes in during the middle of the `complete` phase, all current and future work is cancelled because the graph will need to be updated and we don't want to continue filling out the edges that may no longer be there after the update. We will be able to pick up easily where we left off based on the state of the graph. If any file system changes happen during the `update` phase, we just add that work to the back of the queue. 

There is definitely room for optimization here. I should probably explore using graph state to keep track of `inconsistent` states along with `incomplete` states. That way I could also cancel the `update` phase and start over with minimal work based on the state of the graph. (Since the `update` phase wasn't being cancelled, I wasn't always sending a signal and that's why I needed to check for it). I don't think it's absolutely necessary to have two queues, but I as I was working out the problem it helped me to separate out these two phases and separating out the queues helped me keep track of the flow better. I could see that becoming less necessary if I add more state to the graph like I was mentioning. Then it would be two phases using one queue, but never at the same time. The two phases would act very similarly, one would just be looking `inconsistent` nodes and the other would be looking for `incomplete` nodes. Sound good?

Reply:
This is a good explanation, thanks! Would be great to have this documented somewhere as well at some point (maybe when things are more finalized), perhaps as a comment at the top of the file. Keep me updated on how the investigations you described go. :)